### PR TITLE
Omar/child storage

### LINKED ISSUE #1490: implement store/load of child trie from DB
## Describe the bug

- If `GetStorageChild` or `GetStorageFromChild` is called (in state.StorageState) It fails to load child trie from DB.
- the trie functions `Store` and `WriteDirty` should be updated to write the child tries to the database
- the trie function `Load` should be updated to load the child tries from the database and store them in memory

## To Reproduce

1. Clear the child trie from cache.
2. Try to load the child trie from DB.

## Log output

<details>
  <summary>Log Output</summary>
  
  ```Paste log output here
  paste log output...
  ```
</details> 
</br>

## Specification

- go version:
- gossamer version:
- gossamer commit tag:
- gossamer commit hash:
- operating system:
- additional links:

### PR DESCRIPTION
**Pull request by omar391**
## Changes

- Added test cases for `GetStorageFromChild`
-
-

## Tests

```
go test -timeout 30s -run ^TestStorage_StorageChild$ github.com/ChainSafe/gossamer/dot/state
```

## Issues

- closes #1490

## Primary Reviewer 

@EclesioMeloJunior 
-

### CODE REVIEW

File: dot/state/storage_test.go [hunk 1]
```diff
@@ -23,6 +28,84 @@ func newTestStorageState(t *testing.T) *StorageState {
...
+	err := bs.AddBlock(block0)
+	if err != nil {
+		return err
+	}
+
+	bs.StoreRuntime(block0.Header.Hash(), rt)
+
+	// Create header & blockData for block 1
+	digest := types.NewDigest()
+	err = digest.Add(*types.NewBabeSecondaryPlainPreDigest(0, 1).ToPreRuntimeDigest())
```
Comment (EclesioMeloJunior) [inline dot/state/storage_test.go:84]:
```suggestion
	_, err = digest.Add(*types.NewBabeSecondaryPlainPreDigest(0, 1).ToPreRuntimeDigest())
```

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

File: dot/state/storage_test.go [hunk 2]
```diff
@@ -43,6 +130,79 @@ func TestStorage_StoreAndLoadTrie(t *testing.T) {
...
+			hash = storage.blockState.BestBlockHash()
+		} else {
+			hash = *test.hash
+		}
+
+		stateRoot, err := storage.GetStateRootFromBlock(&hash)
+		require.NoError(t, err)
+
+		item, err := storage.GetStorageFromChild(stateRoot, test.keyChild, test.entry)
+		require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (omar391) [reply dot/state/storage_test.go]: @noot its resolved!

File: dot/state/storage_test.go [hunk 3]
```diff
@@ -43,6 +130,79 @@ func TestStorage_StoreAndLoadTrie(t *testing.T) {
...
+	}{
+		{
+			err:      nil,
+			expect:   uint64(len([]byte(":child_first_value"))),
+			hash:     nil,
+			entry:    []byte(":child_first"),
+			keyChild: []byte(":child_storage_key"),
+		},
+		{
+			err:      nil,
```
[the remark this answers is not in the document]
Comment (omar391) [reply dot/state/storage_test.go]: Done.

File: dot/state/storage_test.go [hunk 4]
```diff
@@ -43,6 +130,79 @@ func TestStorage_StoreAndLoadTrie(t *testing.T) {
...
+
+		stateRoot, err := storage.GetStateRootFromBlock(&hash)
+		require.NoError(t, err)
+
+		item, err := storage.GetStorageFromChild(stateRoot, test.keyChild, test.entry)
+		require.NoError(t, err)
+
+		if item != nil {
+			res = uint64(len(item))
+		}
```
[the remark this answers is not in the document]
Comment (omar391) [reply dot/state/storage_test.go:117]: Done

**Comment by noot** [comment]
going to close this for now as it doesn't address the issue

**Pull request closed without merging**